plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -221,6 +224,45 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
         }
     }
 
+    @Override
+    public JdbcTableHandle getTableHandle(ConnectorSession session, PreparedQuery preparedQuery)
+    {
+        ImmutableList.Builder<JdbcColumnHandle> columns = ImmutableList.builder();
+        try (Connection connection = connectionFactory.openConnection(session);
+                PreparedStatement preparedStatement = queryBuilder.prepareStatement(this, session, connection, preparedQuery)) {
+            ResultSetMetaData metadata = preparedStatement.getMetaData();
+            if (metadata == null) {
+                throw new UnsupportedOperationException(format("Statement not supported: %s. ResultSetMetaData not available", preparedQuery.getQuery()));
+            }
+            for (int column = 1; column <= metadata.getColumnCount(); column++) {
+                JdbcTypeHandle jdbcTypeHandle = new JdbcTypeHandle(
+                        metadata.getColumnType(column),
+                        Optional.ofNullable(metadata.getColumnTypeName(column)),
+                        Optional.of(metadata.getPrecision(column)),
+                        Optional.of(metadata.getScale(column)),
+                        Optional.empty(), // TODO support arrays
+                        Optional.of(metadata.isCaseSensitive(column) ? CASE_SENSITIVE : CASE_INSENSITIVE));
+                Type type = toColumnMapping(session, connection, jdbcTypeHandle)
+                        .orElseThrow(() -> new UnsupportedOperationException("Unsupported type: " + jdbcTypeHandle))
+                        .getType();
+                columns.add(new JdbcColumnHandle(metadata.getColumnName(column), jdbcTypeHandle, type));
+            }
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, "Failed to get table handle for prepared query: " + preparedQuery.getQuery(), e);
```

Comment:
i don't think `preparedQuery.getQuery()` should go into exception message.
it can be super-long.

you can append `firstNonNull(e.getMessage(), e)` to the message

Reply:
OK. In this case, I'll limit the expected message in tests to the common part: `Failed to get table handle for prepared query` to avoid overriding.